File: cpp/network.cpp
#include "network.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const bool shm = false;

Network::Network(int bs, int ifn, int rn, bool it)
{
	board_size = bs;
	input_frame_num = ifn;
	residual_num = rn;
	is_trainable = it;

	system("rm con_*");
	if(shm) {
		system("touch /dev/shm/con_python_enable");
	} else {
		system("touch con_python_enable");
	}
	char str[2048] = "";
	sprintf(str,"python3 -c \"from Network import *; network = Network(board_size = %d, input_frame_num = %d, residual_num = %d, is_trainable=%s)\" &", bs,ifn, rn, it?"True":"False");
	system(str);

	FILE* fp;
	if(shm) {
		while(!(fp = fopen("/dev/shm/con_network_ready","r"))) {
			printf("\rinitializing network");
		}
	} else {
		while(!(fp = fopen("con_network_ready","r"))) {
			printf("\rinitializing network");
		}

	}
	printf("\ninitialized\n");
	fclose(fp);
	if(shm) {
		system("rm /dev/shm/con_network_ready");
	} else {
		system("rm con_network_ready");
	}
}

pair<vector<vector<double> >, double> Network::get_output(vector<Board>& s)
{
	FILE* fp;
	if(shm) {
		fp = fopen("/dev/shm/con_network_input.txt_", "w");
	} else {
		fp = fopen("con_network_input.txt_", "w");
	}
	for(int f=0; f<input_frame_num; f++) {
		for(int i=0; i<board_size; i++) {
			for(int j=0; j<board_size; j++) {
				fprintf(fp, "%d ",s[f][i][j]);
			}
			fprintf(fp, "\n");
		}
		fprintf(fp, "\n");
	}
	fclose(fp);
	if(shm) {
		system("mv /dev/shm/con_network_input.txt_ /dev/shm/con_network_input.txt");
	} else {
		system("mv con_network_input.txt_ con_network_input.txt");
	}

	if(shm) {
		while(!(fp = fopen("/dev/shm/con_network_output.txt","r")));
	} else {
		while(!(fp = fopen("con_network_output.txt","r")));
	}

	auto P = vector<vector<double> > (board_size, vector<double>(board_size, 0.0));
	double V;
	for(int i=0; i<board_size; i++) {
		for(int j=0; j<board_size; j++) {
			fscanf(fp, "%lf", &P[i][j]);
		}
	}
	fscanf(fp, "%lf", &V);
	fclose(fp);
	if(shm) {
		system("rm /dev/shm/con_network_output.txt");
	} else {
		system("rm con_network_output.txt");
	}

	return make_pair(P,V);
}

void Network::train(
	vector<vector<Board> > X,
	vector<vector<vector<double> > > P,
	vector<double> Z,
	int iter,
	int print_iter
)
{
	const int data_num = (int)X.size();
	if((int)P.size() != data_num || (int)Z.size() != data_num) {
		printf("can not train\n");
		return;
	}

	FILE* fp;
	if(shm) {
		fp = fopen("/dev/shm/con_train_data.txt_","w");
	} else {
		fp = fopen("con_train_data.txt_","w");
	}
	fprintf(fp, "%d %d %d\n",data_num, iter, print_iter);
	for(int n=0; n<data_num; n++) {
		for(int f=0; f<input_frame_num; f++) {
			for(int i=0; i<board_size; i++) {
				for(int j=0; j<board_size; j++) {
					fprintf(fp, "%d ", X[n][f][i][j]);
				}
				fprintf(fp, "\n");
			}
			fprintf(fp, "\n");
		}
		for(int i=0; i<board_size; i++) {
			for(int j=0; j<board_size; j++) {
				fprintf(fp, "%lf ", P[n][i][j]);
			}
			fprintf(fp, "\n");
		}
		fprintf(fp, "%lf\n",Z[n]);
		fprintf(fp, "\n");
	}
	fclose(fp);
	if(shm) {
		system("mv /dev/shm/con_train_data.txt_ /dev/shm/con_train_data.txt");
	} else {
		system("mv con_train_data.txt_ con_train_data.txt");
	}
}

bool Network::is_training()
{
	FILE* fp;
	if(shm) {
		fp = fopen("/dev/shm/con_network_training","r");
	} else {
		fp = fopen("con_network_training","r");
	}
	if(fp==NULL) {
		fclose(fp);
		return false;
	}
	fclose(fp);
	return true;
}

void Network::wait_training()
{
	FILE* fp;
	if(shm) {
		while((fp = fopen("/dev/shm/con_network_training","r"))) fclose(fp);
	} else {
		while((fp = fopen("con_network_training","r"))) fclose(fp);
	}
}

File: cpp/Network.py
import tensorflow as tf
import numpy as np
import os

def ConvBlock(x):
    x = tf.layers.conv2d(x, 256, 3, strides=1, padding='same', name='conv')
    x = tf.layers.batch_normalization(x)
    x = tf.nn.relu(x)
    return x

def ResidualBlock(h):
    with tf.variable_scope('res'):
        x = tf.layers.conv2d(h, 256, 3, strides=1, padding='same', name='conv1')
        x = tf.layers.batch_normalization(x, name='bn1')
        x = tf.nn.relu(x)
        x = tf.layers.conv2d(x, 256, 3, strides=1, padding='same', name='conv2')
        x = tf.layers.batch_normalization(x, name='bn2')
        x = x + h
        x = tf.nn.relu(x)
        return x

def PolicyOutputBlock(x, board_size, output_size):
    with tf.variable_scope('policy_output'):
        x = tf.layers.conv2d(x, 2, 1, strides=1, padding='same', name='conv')
        x = tf.layers.batch_normalization(x)
        x = tf.nn.relu(x)
        x = tf.layers.flatten(x)
        x = tf.layers.dense(x, output_size, name='fc')
        x = tf.nn.softmax(x)
        return x

def ValueOutputBlock(x, board_size):
    with tf.variable_scope('value_output'):
        x = tf.layers.conv2d(x, 1, 1, strides=1, padding='same', name='conv')
        
        x = tf.layers.flatten(x)
        x = tf.layers.batch_normalization(x)
        
        x = tf.nn.relu(x)
        x = tf.layers.dense(x, 256, name='fc1')
        x = tf.nn.relu(x)
        x = tf.layers.dense(x, 1, name='fc2')
        x = tf.nn.tanh(x)
        return x



class Network:
    def __init__(self, board_size, input_frame_num, residual_num, is_trainable):
        self.board_size = board_size
        self.input_frame_num = input_frame_num
        self.residual_num = residual_num

        self.g = tf.Graph()
        with self.g.as_default():
            self.X = tf.placeholder(tf.float32, [None, self.board_size, self.board_size, self.input_frame_num])
            self.pi = tf.placeholder(tf.float32, [None, self.board_size**2])
            self.Z = tf.placeholder(tf.float32, [None, 1])

            X = ConvBlock(self.X)
            
            for i in range(self.residual_num):
                with tf.variable_scope('res_%d' % i):
                    X = ResidualBlock(X)
            self.P = PolicyOutputBlock(X, self.board_size, self.board_size**2)
            self.V = ValueOutputBlock(X, self.board_size)
           

            var = tf.trainable_variables()
            theta = tf.reduce_mean(tf.add_n([tf.nn.l2_loss(v) for v in var]) )*1e-4
            self.value_loss = tf.reduce_mean((self.Z - self.V)**2)
            self.policy_loss = tf.reduce_mean( - tf.reduce_sum( tf.multiply(self.pi, tf.log(tf.maximum(self.P,1e-7))), 1, keep_dims=True))
            #self.policy_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=self.pi, logits=self.P))
            self.loss = self.value_loss + self.policy_loss + theta
            
            learning_rate = 0.001 #0.01 # TODO apply learning step anneling
            optimizer = tf.train.MomentumOptimizer(learning_rate, 0.9)
            self.train_model = optimizer.minimize(self.loss)
            self.init = tf.global_variables_initializer()
            
            self.saver = tf.train.Saver()


        self.sess = tf.Session(graph = self.g)
        self.is_trainable = is_trainable
        if is_trainable == True:
            self.sess.run(self.init)
        else:
            self.saver.restore(self.sess,"./AlphaGomoku.ckpt")

        f = open("/dev/shm/con_network_ready","w")
        f.close()
            
        while os.path.isfile("/dev/shm/con_python_enable"):
            if os.path.isfile("/dev/shm/con_network_input.txt"):
                input_data = []
                while len(input_data) == 0:
                    with open("/dev/shm/con_network_input.txt") as f:
                        input_data = f.readlines() 
                os.system("rm /dev/shm/con_network_input.txt")
                input_data = [x.strip() for x in input_data]

                cnt = 0
                s = [0] * input_frame_num
                for fn in range(input_frame_num):
                    f = np.zeros([board_size, board_size])
                    for i in range(board_size):
                        t = input_data[cnt].split(' ')
                        t = [float(x) for x in t]
                        cnt += 1
                        for j in range(board_size):
                            f[i][j] = t[j]
                    s[fn] = f
                    cnt += 1
                s = np.asarray([s])
                s = np.transpose(s, [0, 2, 3, 1])

                P,V = self.get_output(s)
                P = np.reshape(P, [board_size, board_size])
                
                f = open("/dev/shm/con_network_output.txt_","w")
                for i in range(board_size):
                    s = ""
                    for j in range(board_size):
                        s += str(float(P[i][j])) + " "
                    f.write(s+"\n")
                f.write(str(float(V))+"\n")
                f.close()
                os.system("mv /dev/shm/con_network_output.txt_ /dev/shm/con_network_output.txt")

            if os.path.isfile("/dev/shm/con_train_data.txt"):
                f = open("/dev/shm/con_network_training","w")
                f.close()

                input_data = []
                while len(input_data) == 0:
                    with open("/dev/shm/con_train_data.txt") as f:
                        input_data = f.readlines() 
                os.system("rm /dev/shm/con_train_data.txt")
                input_data = [x.strip() for x in input_data]

                data_num = int(input_data[0].split(' ')[0])
                iteration = int(input_data[0].split(' ')[1])
                print_iter = int(input_data[0].split(' ')[2])

                input_data = input_data[1:]

                X = np.zeros([data_num, input_frame_num]).tolist()
                P = [0] * data_num
                Z = [0] * data_num

                cnt = 0
                for n in range(data_num):
                    for fn in range(input_frame_num):
                        X[n][fn] = np.zeros([board_size, board_size])
                        for i in range(board_size):
                            t = input_data[cnt].split(' ')
                            t = [float(x) for x in t]
                            for j in range(board_size):
                                X[n][fn][i][j] = t[j]
                            cnt += 1
                        cnt += 1

                    P[n] = np.zeros([board_size, board_size])
                    for i in range(board_size):
                        t = input_data[cnt].split(' ')
                        t = [float(x) for x in t]
                        for j in range(board_size):
                            P[n][i][j] = t[j]
                        cnt += 1
                    Z[n] = [float(input_data[cnt])]
                    cnt += 1
                    cnt += 1

                X = np.asarray(X)
                P = np.asarray(P)
                Z = np.asarray(Z)

                X = np.transpose(X, [0,2,3,1])
                P = np.reshape(P, [-1,board_size*board_size])

                self.train(X, P, Z, iteration, print_iter)

                os.system("rm /dev/shm/con_network_training")


    def train(self, X_, pi_, Z_, it, prt_it):
        if self.is_trainable:
            for i in range(it): # TODO should change # of iteration steps
                for batch in range(BATCH_SIZE//MINI_BATCH_SIZE):
                    fd = {
                            self.X: X_[batch*MINI_BATCH_SIZE:(batch+1)*MINI_BATCH_SIZE],
                            self.pi: pi_[batch*MINI_BATCH_SIZE:(batch+1)*MINI_BATCH_SIZE],
                            self.Z: Z_[batch*MINI_BATCH_SIZE:(batch+1)*MINI_BATCH_SIZE]
                    }
                    self.sess.run(self.train_model, feed_dict=fd)

                fd = {self.X: X_, self.pi: pi_, self.Z: Z_}
                if i % prt_it == 0:
                    print('======= ' + str(i) + ' =======')
                    l, pl, vl = self.sess.run([self.loss, self.policy_loss, self.value_loss], feed_dict=fd)
                    print('loss: ' , l)
                    print('policy loss: ', pl)
                    print('value loss: ', vl)
                    save_path = self.saver.save(self.sess,"./AlphaGomoku.ckpt")

    def get_output(self, X_):
        return self.sess.run([self.P, self.V], feed_dict={self.X: X_})

